src/coreclr/jit/valuenum.cpp

```diff
@@ -8774,6 +8788,54 @@ void Compiler::fgValueNumberTree(GenTree* tree)
         }
         else // Look up the VNFunc for the node
         {
+            // Check if we can fold GT_ARR_LENGTH on top of a known array (immutable)
+            if (tree->OperIs(GT_ARR_LENGTH))
+            {
+                // Case 1: ARR_LENGTH(FROZEN_OBJ)
+                ValueNum addressVN = tree->gtGetOp1()->gtVNPair.GetLiberal();
+                if (vnStore->IsVNHandle(addressVN) && (vnStore->GetHandleFlags(addressVN) == GTF_ICON_OBJ_HDL))
+                {
+                    size_t handle = vnStore->CoercedConstantValue<size_t>(addressVN);
+                    int    len    = this->info.compCompHnd->getArrayOrStringLength((CORINFO_OBJECT_HANDLE)handle);
+                    if (len >= 0)
+                    {
+                        tree->gtVNPair.SetBoth(vnStore->VNForIntCon(len));
+                        return;
+                    }
+                }
+
+                // Case 2: ARR_LENGTH(static-readonly-field)
+                VNFuncApp funcApp;
+                if (vnStore->GetVNFunc(addressVN, &funcApp) && (funcApp.m_func == VNF_InvariantNonNullLoad))
+                {
+                    ValueNum fieldSeqVN = funcApp.m_args[1];
+                    if ((fieldSeqVN != vnStore->VNForNull()) && (fieldSeqVN != ValueNumStore::NoVN))
+                    {
+                        FieldSeq* fieldSeq = vnStore->FieldSeqVNToFieldSeq(fieldSeqVN);
+                        if (fieldSeq != nullptr)
+                        {
+                            CORINFO_FIELD_HANDLE field = fieldSeq->GetFieldHandle();
+                            if (field != NULL)
+                            {
+                                uint8_t buffer[TARGET_POINTER_SIZE] = {0};
+                                if (this->info.compCompHnd->getReadonlyStaticFieldValue(field, buffer,
+                                                                                        TARGET_POINTER_SIZE, false))
+                                {
+                                    CORINFO_OBJECT_HANDLE objHandle = NULL;
+                                    memcpy(&objHandle, buffer, TARGET_POINTER_SIZE);
```

Comment:
Can you at least add comment that the mismatch in pointer size is intentionally how the JIT-EE interface works?

Reply:
Sure, added